src/v/datalake/coordinator/state_update.cc

```diff
@@ -33,27 +33,29 @@ checked<add_files_update, stm_update_error> add_files_update::build(
       .tp = tp,
       .entries = std::move(entries),
     };
-    if (!update.can_apply(state)) {
-        return stm_update_error::failed;
+    auto allowed = update.can_apply(state);
+    if (allowed.has_error()) {
+        return allowed.error();
     }
     return update;
 }
 
-bool add_files_update::can_apply(const topics_state& state) {
+checked<std::nullopt_t, stm_update_error>
```

Comment:
nit: would be be useful to define a "no error state" in stm_update_error

```
enum class stm_update_error {
    none
    failed,
};

```

that I think removes the need for check<nullopt_t...>

Reply:
I've been preferring to use errors for only error types, and use result/checked to signal expected vs not expected outcomes. I kind of like how error handling is uniform now with `has_error()`, instead of having to guess about optionality or special 0 codes

Reply:
hmm ok, fwiw https://en.cppreference.com/w/cpp/error/error_code/operator_bool also defines using a `none` equivalent with 0 value, so thats somewhat of a standard  approach I guess, anyway I'm not too strong about it.

Reply:
Yea, IMO C++'s "best practices" for error handling are a bit error prone, unless we standardize our codebase to One True Way, it's way to easy to mess up..